app/controllers/api/v2/fact_values_controller.rb

```
@@ -0,0 +1,49 @@
+module Api
+  module V2
+    class FactValuesController < V1::FactValuesController
+      include Api::Version2
+      include Foreman::Controller::SmartProxyAuth
+
+      add_puppetmaster_filters :create
+
+      api :POST, "/fact_values", "Upload facts for a host."
+
+      def index
+        render :text => "hi!", :status => 200
+      end
+
+      def create
+        imported = detect_host_type.importHostAndFacts params.delete("facts")
+        if imported
+          render :text => _("Imported facts"), :status => 200 and return
+        else
+          render :text => _("Failed to import facts"), :status => 400
+        end
+      rescue Exception => e
+        logger.warn "Failed to import facts: #{e}"
+        render :text => _("Failed to import facts: %s") % (e), :status => 400
+      end
+
+      private
+
+      def detect_host_type
+        return Host::Managed if params[:type].blank?
+        if params[:type].constantize.new.kind_of?(Host::Base)
+          logger.debug "Creating host of type: #{params[:type]}"
+          return params[:type].constantize
+        else
+          raise ::Foreman::Exception.new(N_("Invalid type requested for host creation via facts: %s"), params[:type])
+        end
+      rescue => e
+        logger.warn _("A problem occurred when detecting host type: %s") % (e.message)
```

I'm aware this is moved code, but now's as good a time as any to fix it :)  This will return nothing useful, it should raise an exception.